File: go/tee.go
package main

import (
	"os"
)

type TeeWriter struct {
	file   *os.File
	path   string
	append bool
}

func NewTeeWriter(path string, append bool) *TeeWriter {
	flags := os.O_CREATE | os.O_WRONLY
	if append {
		flags |= os.O_APPEND
	} else {
		flags |= os.O_TRUNC
	}

	f, err := os.OpenFile(path, flags, 0644)
	if err != nil {
		return nil
	}

	return &TeeWriter{
		file:   f,
		path:   path,
		append: append,
	}
}

func (tw *TeeWriter) WriteLine(line string) {
	if tw == nil || tw.file == nil {
		return
	}
	tw.file.WriteString(line + "\n")
}

func (tw *TeeWriter) Close() {
	if tw == nil || tw.file == nil {
		return
	}
	tw.file.Close()
	tw.file = nil
}

File: rust/src/tee.rs
use std::fs::{File, OpenOptions};
use std::io::Write;
use std::path::Path;

pub struct TeeWriter {
    file: Option<File>,
}

impl TeeWriter {
    pub fn new(path: &Path, append: bool) -> Option<Self> {
        let mut opts = OpenOptions::new();
        opts.create(true).write(true);
        if append {
            opts.append(true);
        } else {
            opts.truncate(true);
        }
        match opts.open(path) {
            Ok(f) => Some(Self { file: Some(f) }),
            Err(e) => {
                eprintln!("slit: warning: cannot open tee file {:?}: {}", path, e);
                None
            }
        }
    }

    pub fn write_line(&mut self, line: &str) {
        if let Some(ref mut f) = self.file {
            let _ = writeln!(f, "{}", line);
        }
    }

    pub fn close(&mut self) {
        self.file = None;
    }
}

impl Drop for TeeWriter {
    fn drop(&mut self) {
        self.close();
    }
}
